gapic-common/lib/gapic/rest/error.rb

```diff
@@ -37,36 +50,85 @@ class << self
         # it tries to parse and set a detailed message and an error code from
         # from the Google Cloud's response body
         #
+        # @param err [Faraday::Error] the Faraday error to wrap
+        #
+        # @return [ Gapic::Rest::Error]
         def wrap_faraday_error err
           message = err.message
           status_code = err.response_status
+          status = nil
+          details = nil
+          headers = err.response_headers
 
           if err.response_body
-            msg, code = try_parse_from_body err.response_body
+            msg, code, status, details = try_parse_from_body err.response_body
             message = "An error has occurred when making a REST request: #{msg}" unless msg.nil?
             status_code = code unless code.nil?
           end
 
-          Gapic::Rest::Error.new message, status_code
+          Gapic::Rest::Error.new message, status_code, status, details, headers
         end
 
         private
 
         ##
+        # @private
         # Tries to get the error information from the JSON bodies
         #
         # @param body_str [String]
         # @return [Array(String, String)]
         def try_parse_from_body body_str
           body = JSON.parse body_str
-          return [nil, nil] unless body && body["error"].is_a?(Hash)
 
-          message = body["error"]["message"]
-          code = body["error"]["code"]
+          unless body.is_a?(::Hash) && body&.key?("error") && body["error"].is_a?(::Hash)
+            return [nil, nil, nil,
+                    nil]
+          end
+          error = body["error"]
+
+          message = error["message"] if error.key? "message"
+          code = error["code"] if error.key? "code"
+          status = error["status"] if error.key? "status"
 
-          [message, code]
+          details = parse_details error["details"] if error.key? "details"
+
+          [message, code, status, details]
         rescue JSON::ParserError
-          [nil, nil]
+          [nil, nil, nil, nil]
+        end
+
+        ##
+        # @private
+        # Parses the details data, trying to extract the Protobuf.Any objects
+        # from it, if it's an array of hashes. Otherwise returns it as is.
+        #
+        # @param details [Object, nil] the details object
+        #
+        # @return [Object, nil]
+        def parse_details details
+          # For rest errors details will contain json representations of `Protobuf.Any`
+          # decoded into hashes. If it's not an array, of its elements are not hashes,
+          # it's some other case
+          return details unless details.is_a?(::Array) && details.each { |d| d.is_a?(::Hash) }
```

Comment:
I think you want `details.all?` rather than `details.each`. The latter doesn't return anything useful.

(Also suggests there's a test missing here.)

Reply:
:+1:

Reply:
(the test is already there, it tests for the behaviour where this subclause does nothing. I forgot to delete it when refactoring).